src/common/src/util/iter_util.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub trait ZipEqFast<B: IntoIterator>: ExactSizeIterator + Sized
+where
+    B::IntoIter: ExactSizeIterator,
+{
+    /// A specialized version of `zip_eq` for [`ExactSizeIterator`].
+    ///
+    /// It's a separate trait because Rust doesn't support specialization yet.
+    /// See [tracking issue for specialization (RFC 1210)](https://github.com/rust-lang/rust/issues/31844).
+    #[expect(clippy::disallowed_methods)]
+    fn zip_eq_fast(self, other: B) -> impl ExactSizeIterator<Item = (Self::Item, B::Item)> {
+        let other = other.into_iter();
+        assert_eq!(self.len(), other.len());
```

Comment:
By using `debug_assert` here, we will get these two functions interchangeable, but actually, there's no much overhead for always asserting here. Both LGTM.

Reply:
Is it completely useless to make debug mode slightly faster? 🥹